atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/TokenRangeWritesLogger.java

```diff
@@ -0,0 +1,183 @@
+        if (!Sets.symmetricDifference(ranges, newRanges).isEmpty()) {
+            ranges = newRanges;
+            statsPerTable.clear();
+        }
+    }
+
+    <V> void  markWritesForTable(Set<Map.Entry<Cell, V>> entries, TableReference tableRef) {
+        statsPerTable.putIfAbsent(tableRef, new TokenRangeWrites(tableRef, ranges));
+        TokenRangeWrites tokenRangeWrites = statsPerTable.get(tableRef);
+        entries.forEach(entry -> tokenRangeWrites.markWrite(entry.getKey()));
+        tokenRangeWrites.maybeLog();
+    }
+
+    private static final class TokenRangeWrites {
+        private final TableReference tableRef;
+        private final RangeMap<LightweightOppToken, AtomicLong> writesPerRange;
+        private final AtomicLong totalWrites = new AtomicLong(0);
+        private long lastLoggedTime = System.currentTimeMillis();
+
+        private TokenRangeWrites(TableReference tableRef, Set<Range<LightweightOppToken>> ranges) {
+            this.tableRef = tableRef;
+            ImmutableRangeMap.Builder<LightweightOppToken, AtomicLong> builder = ImmutableRangeMap.builder();
+            ranges.forEach(range -> builder.put(range, new AtomicLong(0)));
+            this.writesPerRange = builder.build();
+        }
+
+        private void markWrite(Cell cell) {
+            writesPerRange.get(LightweightOppToken.of(cell)).incrementAndGet();
+            totalWrites.incrementAndGet();
+        }
+
+        private void maybeLog() {
+            if (shouldLog()) {
+                tryLog();
+            }
+        }
+
+        private synchronized void tryLog() {
+            if (shouldLog()) {
+                long numWrites = totalWrites.get();
+                if (!exceedsThresholdOfWrites(numWrites)) {
+                    logNotEnoughWrites(numWrites);
+                    lastLoggedTime = System.currentTimeMillis();
+                    return;
+                }
+                if (distributionNotUniform()) {
+                    logNotUniform();
+                } else {
+                    logUniform(numWrites);
+                }
+                totalWrites.set(0);
+                lastLoggedTime = System.currentTimeMillis();
+            }
+        }
+
+        private boolean shouldLog() {
+            return exceedsThresholdOfWrites(totalWrites.get())
+                    || (System.currentTimeMillis() - lastLoggedTime) > TIME_UNTIL_LOG_MILLIS;
+        }
+
+        private boolean exceedsThresholdOfWrites(long numWrites) {
+            return numWrites > THRESHOLD_WRITES_PER_TABLE;
+        }
+
+        private boolean distributionNotUniform() {
+            List<Long> values = writesPerRange.asMapOfRanges().values().stream().map(AtomicLong::get)
+                    .collect(Collectors.toList());
+            return MathUtils.confidenceDistributionIsNotUniform(values) > CONFIDENCE_FOR_LOGGING;
+        }
+
+        private void logNotEnoughWrites(long numWrites) {
```

Comment:
Think this is probably too verbose, we should just check sufficieintNumberOfWrites && sufficientTimeElapsed and only log if both are true.